system/lib/pthread/thread_mailbox.c

```diff
@@ -112,8 +113,11 @@ void emscripten_thread_mailbox_send(pthread_t thread, task t) {
   notification_state previous =
     atomic_exchange(&thread->mailbox->notification, NOTIFICATION_PENDING);
   if (previous != NOTIFICATION_PENDING) {
-    _emscripten_notify_mailbox(thread,
-                               pthread_self(),
-                               emscripten_main_runtime_thread_id());
+    if (thread->waiting_async) {
+      __builtin_wasm_memory_atomic_notify((int*)thread, -1);
```

Comment:
So I guess this system doesn't actually write  anything to the thread pointer here?

Reply:
I see so we keep waiting in a loop for the next thing.

Reply:
Correct. At no point does the first member of the pthread struct hold anything other than a pointer to itself.